### PR TITLE
The add operator is now capable of creating the path if parents are missing

### PR DESCRIPTION
**Pull request by Chilinot**
This PR also:
* Updates the patch() function to default to `modified = false` for better readability.
* Updates the testsuite to test the local python file instead of the globally installed json_patch module.

### CODE REVIEW

File: test_json_patch.py [hunk 1]
```diff
@@ -2,18 +2,19 @@
...
 __metaclass__ = type
 
 
 sample_json = json.dumps([
-    {"foo": {"one": 1, "two": 2, "three": 3}, "enabled": True},
-    {"bar": {"one": 1, "two": 2, "three": 3}, "enabled": False},
-    {"baz": [{"foo": "apples", "bar": "oranges"},
+    { "foo": {"one": 1, "two": 2, "three": 3}, "enabled": True},
+    { "bar": {"one": 1, "two": 2, "three": 3}, "enabled": False},
+    { "baz": [{"foo": "apples", "bar": "oranges"},
```
Comment (particledecay) [inline test_json_patch.py:13]: Can you turn off whatever auto-formatter is causing this?

**Review by particledecay — CHANGES_REQUESTED** [review]
I'm a little torn on this, as [RFC 6901 actually defines a nonexistent path](https://www.rfc-editor.org/rfc/rfc6901#section-7) as an error condition. That was the general thinking while I was building this out.

However, I also recognize that even by the language of the RFC, handling of the condition is up to the application. This does seem like a useful feature, but it's worth noting that it may break existing implementations out there using this module.

File: test_json_patch.py (hunk 1 above)
Comment (particledecay) [reply test_json_patch.py:13]: @Chilinot?

**Comment by rdxsuper** [comment]
> I'm a little torn on this, as [RFC 6901 actually defines a nonexistent path](https://www.rfc-editor.org/rfc/rfc6901#section-7) as an error condition. That was the general thinking while I was building this out.
> 
> However, I also recognize that even by the language of the RFC, handling of the condition is up to the application. This does seem like a useful feature, but it's worth noting that it may break existing implementations out there using this module.

So as to not break the existing implementations maybe we can add another parameter like say "force" and get the user to set it to 'true' or 'false' depending on whether the user intends to add a non-existing value or patch an existing one. If default is set to 'false' then throws an error upon not finding the property on patch operations. That way we shift the burden on to the user to be aware of the schema.

**Comment by particledecay** [comment]
> add another parameter like say "force"

I like this idea. I'd probably lean toward using `create` rather than `force` however, as I believe the former is more in line with other builtin Ansible modules' behavior (such as `lineinfile`).

So the behavior remains the same for existing deployments, but the addition of `create: yes` tells the module to create the key if it doesn't exist rather than throwing an exception. I'm good with it.

@Chilinot?

**Comment by hydrian** [comment]
Is this still being worked on?  I'm trying to figure out how to add empty object so I can add child objects afterwards because I can't add objects without an existing parent object path.  To me it seems like this should work but it requires a value attribute....
```
- json_patch:
    operations:
      op: add
      path: '/features'
```

**Comment by particledecay** [comment]
> Is this still being worked on?

I haven't heard back from the PR author ever since I responded, so I'm not sure what's going on. I can make the changes myself at this point since it doesn't seem like he's coming back to this.

> I'm trying to figure out how to add empty object so I can add child objects afterwards because I can't add objects without an existing parent object path. To me it seems like this should work but it requires a value attribute....
> 
> ```
> - json_patch:
>     operations:
>       op: add
>       path: '/features'
> ```

Have you tried adding an empty value?

**Comment by hydrian** [comment]
I figured it out but it wasn't very intuitive.

I had to create the parent node with this operation:

`
- name: "Enabling BuildKit support"                                     :                                                             src: '/etc/docker/daemon.json'                                        create: true
    pretty: true                                                          operations:                                                             - op: 'add'                                                             path: '/features'                                                     value: {}
```

I don't think this needs a code change but having an example of creating an empty example object and array would really be helpful.

**Pull request closed without merging**